content/blog/automated-video-transcoding-with-openfaas.md

```diff
@@ -0,0 +1,109 @@
+---
+title: "Automated Video Transcoding with OpenFaaS"
+date: 2018-06-09
+linktitle: Automated Video Transcoding
+draft: false
+---
+
+# Purpose
+
+I have a very large movie collection and we use Plex to keep it organized and easy to manage. That means a lot of storage space is needed for movie files, especially when they are Blue-ray. To save space on disk while maintaining the good quality I built this transcode pipeline to automate the process of adding a new movie to the library. You can find the repo [here](https://github.com/cpitkin/openfaas-transcode)
+
+## Uploader
+
+The companion uploader executable can be found [here](https://github.com/cpitkin/video-upload/)
+
+# Architecture
+
+![architecuture diagram](/img/tp.png)
+
+## Considerations
+
+- Transcoding is a very CPU intensive task. Regardless of how you configure the setup, it is going to take significant CPU and disk i/o resources. That being said you can run this setup on pretty much any hardware but the more resources you give the transcode container the faster it will run. This setup has been tested using an Ubuntu Linux box with a 6 core 3.5 GHz AMD processor.
+
+- We will be uploading raw video to one bucket (transcode) and placing the transcoded video in another bucket (complete). This means you will need enough disk space to accommodate at least your original file size*2. Ideally, for lots of files, you would want at least 500GB of free space.
+
+### Setup
+
+### Prerequisites
+
+You will need a one or two Minio servers setup. In my case, I use two located on two different physical servers but you can easily do it with one.
+
+Buckets:
+
+- transcode
+- complete
+- movies
+- tv
+
+#### Minio config
+
+You should set the Access Key, Secret Key, and webhook values before deploying the container. That will save you the hassle of getting the generated keys and setting the webhook later and needing to restart the container. An example config is below:
+
+```json
+{
+  ...
+  "credential": {
+          "accessKey": "<random_string>",
+          "secretKey": "<random_string>"
+  },
+  "region": "rapture",
+  "browser": "on",
+  "domain": "",
+  "storageclass": {
+          "standard": "",
+          "rrs": ""
+  },
+  "notify": {
+    ...
+    "webhook": {
+      "1": {
+        "enable": true,
+        "endpoint": "http://<IP or URL>:8080/function/transcode-entrypoint"
+      }
+    }
+    ...
+  }
+}
+```
+
+**NOTE:** If you're using the uploader executable mentioned above then the PutObject event trigger will be set for you. If you will need to manually set a PutObject event trigger for the webhook on the transcode bucket.
+
+## OpenFaaS Functions (in order)
```

Comment:
"A picture speaks 1000 words" - put one together with draw.io or similar.

Reply:
There is an architecture reference on line 18